### PR TITLE
Remove some unused or duplicated code

### PR DESCRIPTION
**Pull request by greatroar**
What is the purpose of this change? What does it change?
--------------------------------------------------------

Some unused or duplicated code was removed.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: internal/archiver/tree.go [hunk 1]
```diff
@@ -195,26 +195,13 @@ func (t *Tree) add(fs fs.FS, target, root string, pc []string) error {
 	return nil
 }
 
-func (t Tree) String() string {
```
Comment (MichaelEischer) [inline internal/archiver/tree.go]: This is not dead code. The String method is used via the `debug.Log` calls in `NewTree` (last function in tree)

File: internal/archiver/file_saver.go [hunk 2]
```diff
@@ -53,7 +53,6 @@ type SaveBlobFn func(context.Context, restic.BlobType, *Buffer) FutureBlob
 
 // FileSaver concurrently saves incoming files to the repo.
 type FileSaver struct {
-	fs           fs.FS
```
Comment (MichaelEischer) [inline internal/archiver/file_saver.go]: LGTM

File: internal/archiver/archiver.go [hunk 3]
```diff
@@ -628,42 +628,7 @@ func (arch *Archiver) SaveTree(ctx context.Context, snPath string, atree *Tree,
...
-
-func (fi fileInfoSlice) Swap(i, j int) {
-	fi[i], fi[j] = fi[j], fi[i]
-}
-
-func (fi fileInfoSlice) Less(i, j int) bool {
-	return fi[i].Name() < fi[j].Name()
-}
-
-func readdir(filesystem fs.FS, dir string) ([]os.FileInfo, error) {
```
Comment (MichaelEischer) [inline internal/archiver/archiver.go]: I've always been confused by these two functions. It's good to have this cleaned up

File: internal/archiver/tree.go [hunk 4]
```diff
@@ -195,26 +195,13 @@ func (t *Tree) add(fs fs.FS, target, root string, pc []string) error {
...
-	return s
-}
-
 // unrollTree unrolls the tree so that only leaf nodes have Path set.
 func unrollTree(f fs.FS, t *Tree) error {
 	// if the current tree is a leaf node (Path is set) and has additional
 	// nodes, add the contents of Path to the nodes.
 	if t.Path != "" && len(t.Nodes) > 0 {
 		debug.Log("resolve path %v", t.Path)
-		entries, err := fs.ReadDirNames(f, t.Path)
```
Comment (MichaelEischer) [inline internal/archiver/tree.go]: The `fs.ReadDirNames` and `readdirnames` functions have slightly different semantics. The latter sets `fs.O_NOFOLLOW` whereas the former does not. I'm right now not sure which is the correct behavior.

File: internal/archiver/archiver.go [hunk 5]
```diff
@@ -216,7 +216,7 @@ func (arch *Archiver) SaveDir(ctx context.Context, snPath string, fi os.FileInfo
 		return FutureTree{}, err
 	}
 
-	names, err := readdirnames(arch.FS, dir)
+	names, err := readdirnames(arch.FS, dir, true)
```
Comment (MichaelEischer) [inline internal/archiver/archiver.go]: just by looking at `readdirnames` calls, I can't tell the meaning of the third parameter. It probably better to split this up into `readdirnames` and `readdirnamessorted` (wrapping the first) functions.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]

File: internal/archiver/tree.go (hunk 1 above)
Comment (greatroar) [reply internal/archiver/tree.go]: Ok, restored.

File: internal/archiver/archiver.go (hunk 5 above)
Comment (greatroar) [reply internal/archiver/archiver.go]: In that case, I'd rather just sort on the spot.

File: internal/archiver/tree.go (hunk 4 above)
Comment (greatroar) [reply internal/archiver/tree.go]: I can't find any change in behavior and all the tests pass, but I can solve this by changing readdirnames's signature to

```
// flags are passed to fs.OpenFile. O_RDONLY is implied.
func readdirnames(filesystem fs.FS, dir string, flags int) ([]string, error) {
	f, err := filesystem.OpenFile(dir, flags|fs.O_RDONLY, 0)
	// ...
}
```

This is after I remove the sorted argument in favor of sorting on the spot. Is that ok?

File: internal/archiver/tree.go (hunk 1 above)
Comment (greatroar) [reply internal/archiver/tree.go]: tree_test.go now fails to compile when this method is not present.

**Review by fd0 — APPROVED** [review]
LGTM, thanks a lot for taking the time to clean this up!

**Pull request merged**